File: 14_GridLayoutManager/app/src/main/java/com/adiaz/testinggrids/MainActivity.java
package com.adiaz.testinggrids;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.adiaz.testinggrids.retrofit.LocalSportsRestApi;
import com.adiaz.testinggrids.retrofit.entities.Town;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class MainActivity extends AppCompatActivity implements Callback<List<Town>>, TownsAdapter.ListItemClickListener {

    private static final String TAG = MainActivity.class.getSimpleName();
    public static final String BASE_URL = "https://localsports-web.appspot.com";

    @BindView(R.id.progressBar)
    ProgressBar progressBar;

    @BindView(R.id.recyclerView)
    RecyclerView recyclerView;

    List<Town>mTowns;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(MainActivity.BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .build();
        LocalSportsRestApi localSportsRestApi = retrofit.create(LocalSportsRestApi.class);
        Call<List<Town>> repositoriesCall = localSportsRestApi.queryTowns ();
        repositoriesCall.enqueue(this);
    }

    @Override
    public void onResponse(Call<List<Town>> call, Response<List<Town>> response) {
        Log.d(TAG, "onResponse: townsList -->" + response.body().size());
        progressBar.setVisibility(View.INVISIBLE);
        mTowns = response.body();
        recyclerView.setVisibility(View.VISIBLE);
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        TownsAdapter townsAdapter = new TownsAdapter(mTowns, this);
        recyclerView.setLayoutManager(linearLayoutManager);
        recyclerView.setHasFixedSize(true);
        recyclerView.setAdapter(townsAdapter);
    }

    @Override
    public void onFailure(Call<List<Town>> call, Throwable t) {
        Log.d(TAG, "onFailure: " + t.getMessage());
        progressBar.setVisibility(View.INVISIBLE);
    }

    @Override
    public void onListItemClick(int clickedItemIndex) {
        String toastStr = "Item #" + clickedItemIndex + " clicked.";
        Log.d(TAG, "Toast: " + toastStr);
        Toast.makeText(this, toastStr, Toast.LENGTH_SHORT).show();

        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(this);
        SharedPreferences.Editor editor = sharedPref.edit();
        editor.putString(getString(R.string.town_select), mTowns.get(clickedItemIndex).getName());
        editor.commit();


        Intent intent = new Intent(this, SportsActivity.class);
        startActivity(intent);

    }
}

File: 23_KotlinAndRetrofit2/app/src/main/java/com/adiaz/kotlinandretrofit2/MainActivity.kt
package com.adiaz.kotlinandretrofit2

import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import com.adiaz.kotlinandretrofit2.model.RepositorySearchResult
import kotlinx.android.synthetic.main.activity_main.*
import org.jetbrains.anko.toast
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory

class MainActivity : AppCompatActivity(), Callback<RepositorySearchResult> {

    val TAG = MainActivity::class.java.simpleName

    override fun onFailure(call: Call<RepositorySearchResult>?, t: Throwable?) {
        tvResults.text = "Erroraco ${t?.localizedMessage}"
    }

    override fun onResponse(call: Call<RepositorySearchResult>?, response: Response<RepositorySearchResult>?) {
        tvResults.text = "Resultados: ${response?.body()?.total_count}"
        response?.body()?.items?.forEach {
            tvResults.append("\n\t Name: ${it.name}")
            tvResults.append("\n\t Url: ${it.url}\n")
        }
    }

    private val gitHubSearchApi: GitHubSearchApi

    init {
        val retrofit = Retrofit.Builder()
                .baseUrl("https://api.github.com/")
                .addConverterFactory(MoshiConverterFactory.create())
                .build()
        gitHubSearchApi = retrofit.create(GitHubSearchApi::class.java)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        button.setOnClickListener() {
            val searchRepositories = gitHubSearchApi.searchRepositories(etSearchToken.text.toString())
            searchRepositories.enqueue(this)
        }
    }
}
